### ISSUE TITLE
Observers cannot access native zipkin span instance

### ISSUE THREAD

**Issue by iamtakingiteasy** [openzipkin-contrib/zipkin-go-opentracing#156]
Because opentracing observers are passed private spanImpl wrapper, there is no way for observer to extract native zipkin span instance.

As such, it is impossible to implement e.g. centralized unfinished spans flushing using current state of the library.

Also SpanObserver interface does not provide methods for e.g. log/annotation changes and allow special handling for FinishedWithDuration.

To address both issues, I suggest introducing new interfaces

```go
type ZipkinObserver interface {
	OnStartSpan(sp zipkin.Span, operationName string, options zipkin.SpanOption) (ZipkinSpanObserver, bool)
}

type ZipkinSpanObserver interface {
	OnSetOperationName(operationName string)

	OnSetTag(key, value string)

	OnSetRemoteEndpoint(endpoint *model.Endpoint)

	OnAnnotate(t time.Time, annotation string)

	OnFinishedWithDuration(duration time.Duration)

	OnFinish()
}
```

And new tracer option

```go
WithZipkinObserver(zipkinObserver ZipkinObserver) TracerOption
```

**Comment by jcchavezs** [comment]
Interesting, so your use case is to access the list of unfinished spans, while I think about it I will friendly ping @adriancole to see if there is such use cases somewhere else. 

For now I can mention that in libraries like zipkin-js after some due time we flush unfinished spans to the server with a annotation called `zipkin-js.flush` which was actually inspired in `brave.flush` so I guess brave has this feature. Also this very library has the ability to not to flush spans by default, transferring that responsibility to the user.

**Comment by iamtakingiteasy** [comment]
Yes, I have use-case of flushing unfinished spans, when span is covering long-running operation, such as incoming bidirectional GRPC stream with events going both ways and each possibly causing child spans of execution.

In this case, stream may very well last for many hours or even days, so any sampling information may easily get lost unless span is regularly flushed. In my target use-case I intend flushing on every child span finish and with configurable rates/intervals.

While library itself might benefit from global flush timeout specifically, I still think it would be best to allow for implementation of custom flushing policies using observers as well.
